Make all non-input arguments to functionals part of its options, and pass functional options by const-ref when possible

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29404 Make all non-input arguments to functionals part of its options**
* #29364 Pass F::*FuncOptions instead of torch::nn::*Options to functionals, and make F::*FuncOptions a different class when necessary
* #29360 [C++ API] Add functional overloads for fold, linear, loss, normalization, padding
* #29359 [C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision
* #29358 [C++ API] Add functional overloads for activation, batchnorm, distance, embedding

This PR makes all non-input arguments to functionals part of its options parameters, so that we won't break backward compatibility even if we add or reorder some of the non-input arguments to functionals in the future.

Differential Revision: [D18378526](https://our.internmc.facebook.com/intern/diff/D18378526)

Commit:
Make all non-input arguments to functionals part of its options

kostmo commented:
## CircleCI build failures summary
As of commit c5aab10:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=c5aab10f7c16a8da1e46682a64287bb0dbaa1747).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

pbelevich commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -139,9 +139,8 @@ inline Tensor softmax(const Tensor& input, int64_t dim,
-inline Tensor softmax(const Tensor& input, SoftmaxFuncOptions options,
-                      c10::optional<torch::Dtype> dtype = c10::nullopt) {
-  return detail::softmax(input, options.dim(), dtype);
+inline Tensor softmax(const Tensor& input, SoftmaxFuncOptions options) {
```
functional.py has default values for all options in:
`def softmax(input, dim=None, _stacklevel=3, dtype=None)`
so maybe
```suggestion
inline Tensor softmax(const Tensor& input, SoftmaxFuncOptions options = {}) {
```

yf225 replied:
replied in an earlier comment

pbelevich commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -161,9 +160,8 @@ inline Tensor softmin(const Tensor& input, int64_t dim,
-inline Tensor softmin(const Tensor& input, SoftminFuncOptions options,
-                      c10::optional<torch::Dtype> dtype = c10::nullopt) {
-  return detail::softmin(input, options.dim(), dtype);
+inline Tensor softmin(const Tensor& input, SoftminFuncOptions options) {
```
functional.py has default values for all options in:
`def softmin(input, dim=None, _stacklevel=3, dtype=None)`
so maybe
```suggestion
inline Tensor softmin(const Tensor& input, SoftminFuncOptions options = {}) {
```

yf225 replied:
replied in an earlier comment

pbelevich commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -183,9 +181,8 @@ inline Tensor log_softmax(const Tensor& input, int64_t dim,
-inline Tensor log_softmax(const Tensor& input, LogSoftmaxFuncOptions options,
-                          c10::optional<torch::Dtype> dtype = c10::nullopt) {
-  return detail::log_softmax(input, options.dim(), dtype);
+inline Tensor log_softmax(const Tensor& input, LogSoftmaxFuncOptions options) {
```
functional.py has default values for all options in:
`def log_softmax(input, dim=None, _stacklevel=3, dtype=None)`
so maybe
```suggestion
inline Tensor log_softmax(const Tensor& input, LogSoftmaxFuncOptions options = {}) {
```

yf225 replied:
replied in an earlier comment

pbelevich commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -244,9 +241,8 @@ inline Tensor rrelu(Tensor& input,
-                    bool training = false) {
-  return detail::rrelu(input, options.lower(), options.upper(), options.inplace(), training);
+inline Tensor rrelu(Tensor& input, RReLUFuncOptions options = {}) {
+  return detail::rrelu(input, options.lower(), options.upper(), options.inplace(), options.training());
```
parameters have different order than in functional.py:
`def rrelu(input, lower=1. / 8, upper=1. / 3, training=False, inplace=False)`

yf225 replied:
Thanks for the catch! Fixed

pbelevich commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -427,15 +427,14 @@ inline Tensor max_unpool1d(
 inline Tensor max_unpool1d(const Tensor& input, const Tensor& indices,
-    MaxUnpool1dFuncOptions options,
-    const c10::optional<IntArrayRef>& output_size = c10::nullopt) {
+    MaxUnpool1dFuncOptions options) {
```
I don't mind but I still can't imagine any scenario when pass-by-value is better than pass-by-const-ref

yf225 replied:
Yeah I agreed, I think an alternative is to make all functional options pass-by-const-ref except for `NormalizeFuncOptions`, because we need it to be mutable in order to store output in the `out` Tensor. Currently all functional options are pass-by-value mostly for consistency, and we can revisit this later if perf is a concern.

yf225 replied:
EDIT: I will change all options to pass-by-const-ref except for `NormalizeFuncOptions`, so that we don't need to worry about this later.

yf225 replied:
Fixed

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -82,7 +82,23 @@ struct TORCH_API SoftmaxOptions {
+  SoftmaxFuncOptions(int64_t dim);
+
+  /// Dimension along which Softmax will be computed.
+  TORCH_ARG(int64_t, dim);
```
I believe if we don't pass a non-None `dim` value to Softmax python module and functional, a warning will be shown from `_get_softmax_dim`: 
https://github.com/pytorch/pytorch/blob/c49b324cbf17272d02f3811f517fe82106ce77d4/torch/nn/functional.py#L1193-L1194
https://github.com/pytorch/pytorch/blob/c49b324cbf17272d02f3811f517fe82106ce77d4/torch/nn/functional.py#L1166-L1174
Since `dim=None` is already deprecated in Python frontend, for C++ frontend we should require `dim` to be provided.

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -94,7 +110,23 @@ struct TORCH_API SoftminOptions {
+  SoftminFuncOptions(int64_t dim);
+
+  /// Dimension along which Softmin will be computed.
+  TORCH_ARG(int64_t, dim);
```
replied in an earlier comment

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -106,7 +138,23 @@ struct TORCH_API LogSoftmaxOptions {
+  LogSoftmaxFuncOptions(int64_t dim);
+
+  /// Dimension along which LogSoftmax will be computed.
+  TORCH_ARG(int64_t, dim);
```
replied in an earlier comment

Commit:
Update on "Make all non-input arguments to functionals part of its options"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

pbelevich reviewed (commented):
```
/// Options for the `BatchNorm` module.
struct TORCH_API BatchNormFuncOptions {
```
Should be functional instead of module

Commit:
Update on "Make all non-input arguments to functionals part of its options, and pass functional options by const-ref when possible"

Make all non-input arguments to functionals part of its options

gh-metadata: pytorch pytorch 29404 gh/yf225/41/head

yf225 commented:
> ```
> /// Options for the `BatchNorm` module.
> struct TORCH_API BatchNormFuncOptions {
> ```
> 
> Should be functional instead of module

Thanks for the catch! I will fix this in a follow-up doc cleaning PR
